picky/src/ssh/traits.rs

```diff
@@ -0,0 +1,256 @@
+
+    fn ssh_string_encode(&self, mut stream: impl Write) -> Result<(), Self::Error> {
+        let string_slice = self.as_ref();
+        stream.write_u32::<BigEndian>(string_slice.len() as u32)?;
+        stream.write_all(string_slice.as_bytes())
+    }
+}
+
+pub trait SshStringDecoder {
+    type Error;
+
+    fn ssh_string_decode(&mut self) -> Result<String, Self::Error>;
+}
+
+impl<T> SshStringDecoder for T
+where
+    T: Read,
+{
+    type Error = io::Error;
+
+    fn ssh_string_decode(&mut self) -> Result<String, Self::Error> {
+        let size = self.read_u32::<BigEndian>()? as usize;
+        let mut buffer = vec![0; size];
+        self.read_exact(&mut buffer)?;
+
+        Ok(String::from_utf8_lossy(&buffer).to_string())
+    }
+}
+
+pub trait SshByteArrayEncoder {
+    type Error;
+
+    fn ssh_byte_array_encode(&self, stream: impl Write) -> Result<(), Self::Error>;
+}
+
+impl<T> SshByteArrayEncoder for T
+where
+    T: AsRef<[u8]>,
+{
+    type Error = io::Error;
+
+    fn ssh_byte_array_encode(&self, mut stream: impl Write) -> Result<(), Self::Error> {
+        stream.write_u32::<BigEndian>(self.as_ref().len() as u32)?;
+        stream.write_all(self.as_ref())
+    }
+}
+
+pub trait SshByteArrayDecoder {
+    type Error;
+
+    fn ssh_byte_array_decode(&mut self) -> Result<Vec<u8>, Self::Error>;
+}
+
+impl<T> SshByteArrayDecoder for T
+where
+    T: Read,
+{
+    type Error = io::Error;
+
+    fn ssh_byte_array_decode(&mut self) -> Result<Vec<u8>, Self::Error> {
+        let size = self.read_u32::<BigEndian>()? as usize;
+        let mut buffer = vec![0; size];
+        self.read_exact(&mut buffer)?;
+
+        Ok(buffer)
+    }
+}
+
+pub trait SshMpintEncoder {
+    type Error;
+
+    fn ssh_mpint_encode(&self, stream: impl Write) -> Result<(), Self::Error>;
+}
+
+impl SshMpintEncoder for BigUint {
+    type Error = io::Error;
+
+    fn ssh_mpint_encode(&self, mut stream: impl Write) -> Result<(), Self::Error> {
+        let data = self.to_bytes_be();
+        let size = data.len() as u32;
+        // If the most significant bit would be set for
+        // a positive number, the number MUST be preceded by a zero byte.
+        if size > 0 && data[0] & 0b10000000 != 0 {
+            stream.write_u32::<BigEndian>(size + 1)?;
+            stream.write_u8(0)?;
+        } else {
+            stream.write_u32::<BigEndian>(size)?;
+        }
+        stream.write_all(&data)
+    }
+}
+
+pub trait SshMpintDecoder {
+    type Error;
+
+    fn ssh_mpint_decode(&mut self) -> Result<BigUint, Self::Error>;
+}
+
+impl<T> SshMpintDecoder for T
+where
+    T: Read,
+{
+    type Error = io::Error;
+
+    fn ssh_mpint_decode(&mut self) -> Result<BigUint, Self::Error> {
+        let size = self.read_u32::<BigEndian>()? as usize;
+        let mut buffer = vec![0; size];
+        self.read_exact(&mut buffer)?;
+
+        if buffer[0] == 0 {
+            buffer.remove(0);
+        }
+
+        Ok(BigUint::from_bytes_be(&buffer))
+    }
+}
```

Comment:
I did it due to lack of specialization as you pointed and to support encoding values by reference(e.g `&[u8]` and `Vec<u8>` would fit `SshByteArrayEncoder<T: AsRef<u8>>`, splitting SshParser to SshEncode and SshDecode will fix the problem). Also, I think It more readable to have something like `let data =  buffer.ssh_byte_array_decode()` than `let data: Vec<u8> = buffer.ssh_decode()`, the first variant describes intention more clear.

Reply:
> it requires to clone everything

Yes, you need to separate at least encoding from decoding because you can't exactly decode to a slice in an owned fashion.

> Also, I think It more readable to have something like `let data = buffer.ssh_byte_array_decode()` than `let data: Vec<u8> = buffer.ssh_decode()`, the first variant describes intention more clear.

I see your point.

For this we might in fact get simpler and clearer code by just using a function generic over `T: Read` or `&mut impl Read` :thinking: 
The problem might be that these doesn't really need to be traits at all? `SshParser` _is_ useful as a trait (especially the generic implementation `Vec<T>`), but the others feel more like "traits extensions" to `Read` now that I think about it.

What about either:
- using plain functions instead,
- or move all of these in a single new trait extension (`SshReadExt` and `SshWriteExt`).

On the naming, what do you think of:
- `write_ssh_bytes`
- `write_ssh_mpint`
- `read_ssh_bytes`
- `read_ssh_mpint`

For the "bytes" variants, I definitely agree here, we want to use `write_all` and `read_exact` for the specific cases of `&[u8]` and `Vec<u8>` (I understand we can't do that easily using `SshParser` for `Vec<u8>` because we have the implementation for `Vec<T>` that is depending on it).
`SshParser` might be implemented on `BigUint`, but for the sake of clarity it's maybe a good idea not to, I'll leave that to your preference.
However, I think `SshParser` could be implemented on `String`, I don't see the benefit to do otherwise for this one (ssh doc uses `string` as well).

What do you think? :open_mouth: 

(to be clear, I maintain my request about module splitting)

Reply:
I like the idea with `SshReadExt` and `SshWriteExt`(personally It reminds me the `AsyncReadExt` and `AsyncWriteExt` traits from `tokio` :slightly_smiling_face: ), and the functions naming. Now, I also feel that the traits are more like extensions. 

> I think SshParser could be implemented on String, I don't see the benefit to doing otherwise for this one

I think the benefit would be keeping everything _consistent_. `string` in _SSH_ docs is one of the data representations as well as `Mpint`, `bytes` etc. Maybe, it's better to handle `string` in the same as `bytes` and `Mpint`. 
Also, when I was writing the trait, I thought that the `SshParser` trait would be used for complex types (like `CriticalOptions`), and other traits would be used for decoding/encoding basic data types.

Reply:
> I like the idea with `SshReadExt` and `SshWriteExt`(personally It reminds me the `AsyncReadExt` and `AsyncWriteExt` traits from `tokio` slightly_smiling_face ), 

That's exactly it :laughing: 

> > I think SshParser could be implemented on String, I don't see the benefit to doing otherwise for this one
> 
> I think the benefit would be keeping everything _consistent_. `string` in _SSH_ docs is one of the data representations as well as `Mpint`, `bytes` etc. Maybe, it's better to handle `string` in the same as `bytes` and `Mpint`. Also, when I was writing the traits, I thought that the `SshParser` trait would be used for complex types (like `CriticalOption`), and other traits would be used for decoding/encoding basic data types.

I see your point here, sounds good to me! :slightly_smiling_face:

Reply:
So, following this idea, maybe you could go for a `SshComplexType` trait instead. Maybe that's all it takes to make the intention crystal clear! What do you think?

Reply:
`SshComplexType` trait instead of `SshParser` looks awesome!